### PR TITLE
feat: add opentelemetry tracing

### PR DESCRIPTION
**Pull request by bloodearnest**
This adds opentelemetry tracing to running jobs.

A new jobrunner/tracing.py module handles most of the work, and we use
it to trace state changes between the fine grained StatusCode states
added previously.  We store as many attributes on each span as possible.

This includes a migration to add somewhere to store a job's trace
context and last changed timestamp.

We also emit "marker spans" that indicate that a the job has entered
a certain state, but not yet exited it yet. This is intedended to be
useful for ongoing jobs.

Also, we add a final "RUN" span once the job has completed that serves
as coarser grained span to use for graphing whole job metics.

### CODE REVIEW

File: jobrunner/create_or_update_jobs.py [hunk 1]
```diff
@@ -226,6 +226,8 @@ def recursively_build_jobs(jobs_by_action, job_request, pipeline_config, action)
         job_request_id=job_request.id,
         state=State.PENDING,
         status_code=StatusCode.CREATED,
+        # time in ns
```
Comment (madwort) [inline jobrunner/create_or_update_jobs.py:229]:
```suggestion
        # time in nanoseconds
```

File: jobrunner/create_or_update_jobs.py [hunk 2]
```diff
@@ -331,6 +334,8 @@ def create_failed_job(job_request, exception):
         job_request_id=job_request.id,
         state=state,
         status_code=code,
+        # time in ns
```
Comment (madwort) [inline jobrunner/create_or_update_jobs.py:337]:
```suggestion
        # time in nanoseconds
```

File: jobrunner/create_or_update_jobs.py (hunk 1 above)
Comment (madwort) [reply jobrunner/create_or_update_jobs.py:229]: the nittiest of nits

File: jobrunner/models.py [hunk 3]
```diff
@@ -255,7 +275,7 @@ def slug(self):
         Use a human-readable slug rather than just an opaque ID to identify jobs in
         order to make debugging easier
         """
-        return slugify(f"{self.project}-{self.action}-{self.id}")
+        return slugify(f"{self.workspace}-{self.action}-{self.id}")
```
Comment (madwort) [inline jobrunner/models.py:278]: ?

File: jobrunner/run.py [hunk 4]
```diff
@@ -469,17 +477,36 @@ def set_code(job, code, message, error=None, **attrs):
...
 
+        # use higher precision timestamp for state change time
+        job.status_code_updated_at = timestamp_ns
         update_job(job)
+
+        if code in FINAL_STATUS_CODES:
+            # transitioning to a final state, so just record that state
+            tracing.finish_current_state(
+                job, timestamp_ns, final=True, error=error, message=message, **attrs
+            )
```
Comment (madwort) [inline jobrunner/run.py:503]: would be nice to pull this into a special function to clarify purpose

File: jobrunner/tracing.py [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+    emitting so few spans that it is easier just to keep it simpler, and it
+    makes testing easier.
+    """
+    if "OTEL_EXPORTER_OTLP_HEADERS" in os.environ:
+        if "OTEL_SERVICE_NAME" not in os.environ:
+            raise Exception(
+                "OTEL_EXPORTER_OTLP_HEADERS is configured, but missing OTEL_SERVICE_NAME"
+            )
+        if "OTEL_EXPORTER_OTLP_ENDPOINT" not in os.environ:
+            os.environ["OTEL_EXPORTER_OTLP_ENDPOINT"] = "https://api.honeycomb.io"
```
Comment (madwort) [inline jobrunner/tracing.py:39]: open question - should this default to functioning (honeycomb url) or safe (invalid url)? for TPP we're confident that if it defaults to honeycomb the firewall will block it, but is that an assumption we can lean on in future?

File: jobrunner/tracing.py [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+    return {k: "" if v is None else v for k, v in attrs.items()}
+
+
+def initialise_trace(job):
+    """Initialise the trace for this job by creating a root span.
+
+    We store the serialised trace context in the db, so we can reuse it for
+    later spans.
+
+    We create a root span, which is a requirement in OTEl. For this reason we
```
Comment (madwort) [inline jobrunner/tracing.py:91]:
```suggestion
    We create a root span, which is a requirement in OTel. For this reason we
```
nit

File: jobrunner/tracing.py [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+    assert job.status_code is not None, "job has no initial StatusCode"
+    assert (
+        job.status_code_updated_at is not None
+    ), "job has no initial status_code_updated_at"
+
+    job.trace_context = {}
+    tracer = trace.get_tracer("jobs")
+    attrs = trace_attributes(job)
+
+    # convert created_at to ns
```
Comment (madwort) [inline jobrunner/tracing.py:104]:
```suggestion
    # convert created_at to nanoseconds
```

File: jobrunner/tracing.py [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+    start_time = int(job.created_at * 1e9)
+
+    # TraceContextTextMapPropagator only works with the current span, sadlmy
```
Comment (madwort) [inline jobrunner/tracing.py:107]:
```suggestion
    # TraceContextTextMapPropagator only works with the current span, sadly
```

File: jobrunner/tracing.py [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+        if final:
+            # final states have no duration, so make last for 1 sec, just act
+            # as a marker
+            end_time = int(end_time + 1e9)
+
+        record_job_span(job, name, start_time, end_time, error, **attrs)
+
+        if final:
+            # record a full span for the entire run
+            # trace vanity: have the job start 1us before the actualy job, so
```
Comment (madwort) [inline jobrunner/tracing.py:136]:
```suggestion
            # trace vanity: have the job start 1us before the actual job, so
```

File: jobrunner/tracing.py [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+    # trace the initial job state trace
+    start_new_state(job, job.status_code_updated_at)
+
+
+def finish_current_state(job, timestamp_ns, final=False, error=None, **attrs):
+    """Record a span representing the state we've just exited."""
+    try:
+        name = job.status_code.name
+        # Note: this *must* be timestamp as integer nanoseconds
+        start_time = job.status_code_updated_at
```
Comment (madwort) [inline jobrunner/tracing.py:123]: wondering whether renaming this var to `job.status_code_updated_at_ns` could remove some of these comments?

File: jobrunner/tracing.py [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+            # final states have no duration, so make last for 1 sec, just act
+            # as a marker
+            end_time = int(end_time + 1e9)
+
+        record_job_span(job, name, start_time, end_time, error, **attrs)
+
+        if final:
+            # record a full span for the entire run
+            # trace vanity: have the job start 1us before the actualy job, so
+            # it shows up first in the trace
```
Comment (madwort) [inline jobrunner/tracing.py:137]: YES!!

File: jobrunner/tracing.py [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+        # make sure trace failures do not error the job
+        logger.exception(f"failed to trace state for {job.id}")
+
+
+def start_new_state(job, timestamp_ns, error=None, **attrs):
+    """Record a marker span to say that we've entered a new state."""
+    try:
+        name = f"ENTER {job.status_code.name}"
+        start_time = timestamp_ns
+        # fix the time for these synthenic marker events at one second
```
Comment (madwort) [inline jobrunner/tracing.py:151]:
```suggestion
        # fix the time for these synthetic marker events at one second
```

File: jobrunner/tracing.py (hunk 10 above)
Comment (madwort) [reply jobrunner/tracing.py:123]: hmmm, some other things are nanoseconds but not `_ns~`, maybe just leave it as-is

File: tests/test_create_or_update_jobs.py [hunk 13]
```diff
@@ -21,7 +21,7 @@ def disable_github_org_checking(monkeypatch):
 
 
 # Basic smoketest to test the full execution path
-def test_create_or_update_jobs(tmp_work_dir):
+def test_create_or_update_jobs(tmp_work_dir, db):
```
Comment (madwort) [inline tests/test_create_or_update_jobs.py:24]: `db` seems unrelated?

**Review by madwort — CHANGES_REQUESTED** [review]
looks good, just a few nits

File: tests/test_create_or_update_jobs.py (hunk 13 above)
Comment (bloodearnest) [reply tests/test_create_or_update_jobs.py:24]: We now query the db for job_request metadata, where as before we just used the `job` instance that. Given that, this change avoid using the default db and gives it its own isolated db for hygiene.

File: jobrunner/tracing.py (hunk 5 above)
Comment (bloodearnest) [reply jobrunner/tracing.py:39]: It's a fair question.

I think given that we're happy for any such data to live in honeycomb eventually, I think it's ok?

It makes configuring local runs to emit to honeycomb a bit easier, is the main benefit.

Comment (madwort) [reply jobrunner/tracing.py:39]: yep ok

**Review by madwort — APPROVED** [review]
